templates/cc/string.go

```diff
@@ -135,12 +144,39 @@ const strTpl = `
 		}
 		*/}}
 	{{ else if $r.GetHostname }}
-		{{ unimplemented }}
-		{{/* TODO(akonradi) implement hostname constraints
-		if err := m._validateHostname({{ accessor . }}); err != nil {
-			return {{ errCause . "err" "value must be a valid hostname" }}
+	{
+		// Take a copy so we can lowercase it.
+		std::string value = {{ accessor . }};
+		std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c){ return std::tolower(c); });
+
+		if (value.length() > 253) {
+			{{ err . "value must be a hostname, and hostname cannot exceed 253 characters" }}
 		}
-		*/}}
+
+		const std::regex dot_regex{"\\."};
```

Comment:
It's worth noting that all of these checks can be done with a regex. Is the reason to do them as separate checks for better error messages?

Reply:
I wanted to keep the implementation as close to the same as golang implementation. I'm happy to change both of them but think it's important to keep the same internal validation between both. So config that's valid in one language doesn't become invalid in the next.

As for why it's not done with a regex. I think the main part is catching the: `-` beginning/end. I don't know of an efficient way you could do that without splitting the string, and if you're splitting the string, it's quicker to just take the one if branch.